### PR TITLE
Kafka example

### PR DESCRIPTION
**Pull request by zbendhiba**
I need to finish the Readme part. And put difficulties to get steps
But pushing the Kafka example, as it is fully working

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -807,15 +807,64 @@ When you have time, we invite you to take a look at:
...
+* Serialize Data
+* Create REST API
+
+
+We will also experience the joy of using [dev services](https://quarkus.io/guides/dev-services) in the development mode. No need to install any kafka server or database.
+
+---
+**🚀NOTE**
+
+This section requires having Docker installed in order to provition a Kafka and Postgresql servers.
```
Comment (aldettinger) [inline README.md]: Provision a Kafka ?

Comment (aldettinger) [reply README.md]: Also, shouldn't we pre-pull some docker images in the prerequisites section ?

File: README.md [hunk 2]
```diff
@@ -807,15 +807,64 @@ When you have time, we invite you to take a look at:
...
+
+In case, you don't have a proper Docker license:
+* Use H2 Databse instead of Postgresql
+* Ask us during the session to provide you a kafka instance in the Cloud.
+
+---
+
+### Introducing the example
+
+This example will illustrate a application that will receive and process coffee orders. The application is designed around Event Driven Architecture and uses Kafka to communicate with external systems. The orders are stored withing a database. The delivery team is notified on slack to go check the order. The delivery team can access information about orders to deliver using an API.
```
Comment (aldettinger) [inline README.md]: an application that ?

Comment (aldettinger) [reply README.md]: withing => within ?

File: README.md [hunk 3]
```diff
@@ -807,15 +807,64 @@ When you have time, we invite you to take a look at:
...
+
+
+* insert the orders within a Database
+* notify the delivery company using a specific API (This one will be provided during the session)
+
+The application is designed around Event Driven Architecture and uses Kafka to handle those messages:
+* The beer orders to be processed are in a kafka topic named `orders`
```
Comment (aldettinger) [inline README.md]: Is there beer amongst with coffee orders too ? Or is it just coffee ?

File: README.md [hunk 4]
```diff
@@ -807,15 +807,64 @@ When you have time, we invite you to take a look at:
...
+* The notifications to the delivery company are in a kafka topic named `deliveries`
+
+The application that create orders and notify delivery teams are external to our ecosystem.
+
+![](images/CQ-workshop1.png)
+
```
Comment (aldettinger) [inline README.md:704]: I like the "Kaffee" name :)

File: README.md [hunk 5]
```diff
@@ -807,15 +807,64 @@ When you have time, we invite you to take a look at:
...
 
 ## Bonus A - Camel Quarkus and the Kamelets
```
Comment (aldettinger) [inline README.md]: At end of part-7 native mode: The last sentence "That's no big deal as we'll prefer to use the DEV and JVM mode for the rest of the workshop." is no more true now that we JVM + Native at the end ?

File: part-7-kafka/pom.xml [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.camel.quarkus</groupId>
+      <artifactId>camel-quarkus-kafka</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.camel.quarkus</groupId>
+      <artifactId>camel-quarkus-http</artifactId>
+    </dependency>
+    <dependency>
```
Comment (aldettinger) [inline part-7-kafka/pom.xml]: Could we remove the double depenceny to camel-quarkus-htttp ?

File: bonus-a-cloud/.mvn/wrapper/maven-wrapper.properties [hunk 7]
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip
```
Comment (aldettinger) [inline bonus-a-cloud/.mvn/wrapper/maven-wrapper.properties]: I see other sections have 3.8.1, maybe we should align ?

File: pom.xml [hunk 8]
```diff
@@ -86,5 +86,7 @@
         <module>part-4-routes</module>
         <module>part-5-extensions</module>
         <module>part-6-eips</module>
```
Comment (aldettinger) [inline pom.xml]: Could you please test the maven dependencies pre-download section ? I think this PR will break it.

File: README.md (hunk 1 above)
Comment (zbendhiba) [reply README.md]: yes and yes

File: README.md (hunk 3 above)
Comment (zbendhiba) [reply README.md]: just coffee I changed my mind

File: bonus-a-cloud/.mvn/wrapper/maven-wrapper.properties (hunk 7 above)
Comment (zbendhiba) [reply bonus-a-cloud/.mvn/wrapper/maven-wrapper.properties]: Yes, Maybe one of the projects have been generated after the quarkus maven wrapper has been updated

File: README.md (hunk 5 above)
Comment (zbendhiba) [reply README.md]: Didn't understand this comment

Comment (aldettinger) [reply README.md]: It's like in part-7, the last part of the workshop. We say, no big deal as we'll use the DEV mode from now on. Actually,the DEV mode has already been used before in other parts ? Now that this PR is re-ordering parts

Comment (zbendhiba) [reply README.md]: okay !! Understood now

**Comment by zbendhiba** [comment]
@aldettinger  I have 2 TODOs : 

- add the right images of PostgreSQL and Strimzi
- review Quarkus version of wrapper

**Comment by zbendhiba** [comment]
Could you review this, so that I could move on with the renaming thing and create a new PR

File: README.md [hunk 9]
```diff
@@ -807,20 +623,243 @@ When you have time, we invite you to take a look at:
  + [The dynamic router EIP](https://camel.apache.org/components/latest/eips/dynamicRouter-eip.html)
  + [The kamelet EIP](https://camel.apache.org/components/latest/eips/kamelet-eip.html)
 
-## Part 7 - Camel Quarkus and Kafka
+## Part 5 - Camel Quarkus et Kafka
```
Comment (aldettinger) [inline README.md]: "Camel Quarkus and Kafka" ? It looks liek this PR has changed it back to "et" if I read well.

Comment (zbendhiba) [reply README.md]: Nice catch. Updated

File: README.md [hunk 10]
```diff
@@ -46,7 +36,7 @@ It is strongly advised to install as much as possible prior to the workshop.
  * JDK 11 installed
  * Maven >= 3.6.3 advised, possibly with JAVA_HOME configured appropriately
  * A favorite HTTP client like curl
- * Docker >= 1.13.1 installed (if docker is missing, that's no big deal, you may only miss part-3 and part-7)
+ * Docker >= 1.13.1 installed (if docker is missing, that's no big deal, you may only miss part-5 and some concepts in part-5)
```
Comment (aldettinger) [inline README.md]: Not sure about this part. After the renaming part-5 is kafka and part-7 is native-mode.
Is the intent to say "you may only miss part-7 and some concepts in part-5" ?

File: README.md [hunk 11]
```diff
@@ -809,20 +620,367 @@ When you have time, we invite you to take a look at:
...
+The application that create orders and notify delivery teams are external to our ecosystem.
+
+![](resources/CQ-workshop1.png)
+
+Let's navigate to the part 5 folder.
+
+### Dev Services
+Quarkus supports the automatic provisioning of unconfigured services in development and test mode. For more information, you can check the [official Dev Services overview](https://quarkus.io/guides/dev-services).
+As we are using kafka and PostgreSQL, without any configuration, Quarkus dev services will start automatically Strimzi and PostgreSQL. 
+Start the app in dev mode. Notice that those 2 servers have been started.
```
Comment (aldettinger) [inline README.md:658]: I see warnings here, not sure this is an issue:
```
2022-01-21 12:48:38,521 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (JPA Startup Thread: <default>) SQL Warning Code: 0, SQLState: 00000

2022-01-21 12:48:38,522 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (JPA Startup Thread: <default>) table "camel_messageprocessed" does not exist, skipping
2022-01-21 12:48:38,523 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (JPA Startup Thread: <default>) SQL Warning Code: 0, SQLState: 00000
2022-01-21 12:48:38,523 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (JPA Startup Thread: <default>) table "coffeeorder" does not exist, skipping
2022-01-21 12:48:38,523 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (JPA Startup Thread: <default>) SQL Warning Code: 0, SQLState: 00000
2022-01-21 12:48:38,524 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (JPA Startup Thread: <default>) sequence "hibernate_sequence" does not exist, skipping
```

```
2022-01-21 12:48:39,038 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'sasl.kerberos.ticket.renew.window.factor' was supplied but isn't a known config.
2022-01-21 12:48:39,038 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'sasl.kerberos.kinit.cmd' was supplied but isn't a known config.
2022-01-21 12:48:39,039 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'specific.avro.reader' was supplied but isn't a known config.
2022-01-21 12:48:39,039 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'sasl.kerberos.ticket.renew.jitter' was supplied but isn't a known config.
2022-01-21 12:48:39,039 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'ssl.trustmanager.algorithm' was supplied but isn't a known config.
2022-01-21 12:48:39,039 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'ssl.keystore.type' was supplied but isn't a known config.
2022-01-21 12:48:39,039 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'sasl.kerberos.min.time.before.relogin' was supplied but isn't a known config.
2022-01-21 12:48:39,040 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'ssl.protocol' was supplied but isn't a known config.
2022-01-21 12:48:39,040 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'ssl.enabled.protocols' was supplied but isn't a known config.
2022-01-21 12:48:39,040 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'ssl.truststore.type' was supplied but isn't a known config.
2022-01-21 12:48:39,040 WARN  [org.apa.kaf.cli.con.ConsumerConfig] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) The configuration 'ssl.keymanager.algorithm' was supplied but isn't a known config.
2022-01-21 12:48:39,043 INFO  [org.apa.cam.com.kaf.con.sup.ResumeStrategyFactory] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) Using NO-OP resume strategy
2022-01-21 12:48:39,043 INFO  [org.apa.cam.com.kaf.KafkaFetchRecords] (Camel (camel-1) thread #1 - KafkaConsumer[orders]) Subscribing orders-Thread 0 to topic orders
2022-01-21 12:49:11,950 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:49:11,951 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:49:12,947 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:49:12,948 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:49:17,947 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:49:17,948 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:49:37,951 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:49:37,952 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:49:38,951 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:49:38,952 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:49:52,954 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:49:52,955 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:50:28,999 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:50:29,000 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:50:29,227 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {s}->https://random-data-api.com:443: The target server failed to respond
2022-01-21 12:50:29,228 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {s}->https://random-data-api.com:443
2022-01-21 12:50:41,068 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:50:41,069 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:50:41,295 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {s}->https://random-data-api.com:443: The target server failed to respond
2022-01-21 12:50:41,296 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {s}->https://random-data-api.com:443
2022-01-21 12:50:41,961 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:50:41,962 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:51:19,963 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:51:19,965 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:51:20,220 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {s}->https://random-data-api.com:443: The target server failed to respond
2022-01-21 12:51:20,221 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {s}->https://random-data-api.com:443
2022-01-21 12:51:34,965 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://random-data-api.com:80: The target server failed to respond
2022-01-21 12:51:34,966 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {}->http://random-data-api.com:80
2022-01-21 12:51:35,190 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {s}->https://random-data-api.com:443: The target server failed to respond
2022-01-21 12:51:35,192 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {s}->https://random-data-api.com:443
2022-01-21 12:52:12,174 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {s}->https://random-data-api.com:443: The target server failed to respond
2022-01-21 12:52:12,175 INFO  [org.apa.htt.imp.exe.RetryExec] (Camel (camel-1) thread #2 - timer://create-random-coffee-orders) Retrying request to {s}->https://random-data-api.com:443
```

File: README.md [hunk 12]
```diff
@@ -809,20 +620,367 @@ When you have time, we invite you to take a look at:
...
+
+    @Override
+    public void configure() throws Exception {
+        from("timer:create-random-coffee-orders")
+                .to("http:{{random-coffee-api}}");
+    }
+}
+```
+`random-coffee-api` value is described in the [application.properties configuration file](part-5-kafka/src/main/resources/application.properties). It contains the URI to the random coffee API.
+The response from the API is in json format. We use the [Jackson is a Data Format](https://camel.apache.org/components/latest/dataformats/json-jackson-dataformat.html) to convert from Json to [Coffee Java class](part-5-kafka/src/main/java/org/acme/Coffee.java).
```
Comment (aldettinger) [inline README.md]: Maybe we replace "Jackson is a Data Format" -> "Jackson Data Format" ?

File: README.md [hunk 13]
```diff
@@ -809,20 +620,367 @@ When you have time, we invite you to take a look at:
...
+
+```
+    .unmarshal().json(Coffee.class)
+```
+
+Then we use [bean component](https://camel.apache.org/components/latest/bean-component.html) to transform the body to a [Coffee Order](part-5-kafka/src/main/java/org/acme/CoffeeOrder.java). Note that we do specify the method everytime the bean class contains many methods.
+```
```
Comment (aldettinger) [inline README.md:683]: "Note that we do specify the method everytime the bean class contains many methods."
=>
"Note that we explicitly specify the method name `generateOrder` to select amongth the `myBean` methods."

File: README.md (hunk 11 above)
Comment (zbendhiba) [reply README.md:658]: Can you try again? The API is accessible right now for me

File: README.md [hunk 14]
```diff
@@ -116,7 +108,7 @@ cd "${CQ_WORKSHOP_DIRECTORY}/camel-quarkus-workshop"
 mvn quarkus:go-offline -fae
 mvn dependency:go-offline -fae
 mvn clean package -fae
-cd part-2-jvm-mode
+cd part-6-jvm-mode
```
Comment (aldettinger) [inline README.md]: Here, if we do `cd part-5-kafka/` instead. Then the right docker images should be pulled for free ? Might be another PR.

Comment (zbendhiba) [reply README.md]: yes

**Comment by zbendhiba** [comment]
@aldettinger pushed the changes

**Comment by aldettinger** [comment]
@zbendhiba I'm merging now as this PR is blocking other updates to the pre-requisites section that workshop users might see a week ahead.

You are welcome to fix the remaining issues in other PRs.
